### PR TITLE
Add Generic Downsampling to Pipeline

### PR DESCRIPTION
**Pull request by ssanderson**
Adds a new ``downsample`` method to all computable terms.  Computable
terms (Filters, Factors, and Classifiers) can be downsampled to yearly,
quarterly, monthly, or weekly frequency.

The result of ``term.downsample`` is a new term of the same
family (Filter/Factor/Classifier) as ``term``.  The downsampled term
computes by delegating to the original term; repeatedly calling its
``compute`` method with length-1 date ranges.

Downsampled terms take advantage of a new ``compute_extra_rows`` Term
method, which allows terms to dynamically request that additional extra
rows of themselves be computed based on the dates for which they're
being computed.  This ensures, for example, that a monthly-downsampled
term always computes at the start of a month, even when a
naively-calculated pipeline window would end in the middle of the month.

### CODE REVIEW

File: zipline/pipeline/graph.py [hunk 1]
```diff
@@ -34,21 +33,23 @@ class TermGraph(DiGraph):
...
     """
     def __init__(self, terms):
-        super(TermGraph, self).__init__(self)
+        super(TermGraph, self).__init__()
 
         self._frozen = False
         parents = set()
-        for term in itervalues(terms):
-            self._add_to_graph(term, parents, extra_rows=0)
+        for name, term in iteritems(terms):
```
Comment (twiecki) [inline zipline/pipeline/graph.py]: Seems like `name` doesn't get used.

File: zipline/pipeline/term.py [hunk 2]
```diff
@@ -516,6 +585,31 @@ def postprocess(self, data):
...
+
+    def downsample(self, frequency):
+        """
+        Make a term that computes from ``self`` at lower-than-daily frequency.
+
+        Parameters
+        ----------
+        frequency : str, {'Y', 'Q', 'M', 'W'}
+            A string indicating the desired sampling rate.
+            'Y' -> sample on the first trading day of each calendar year
```
Comment (twiecki) [inline zipline/pipeline/term.py]: Often you also want the last trading day of a YQMW. Not saying it has to be part of this PR but preferably the syntax does not exclude that from getting implemented.

Comment (llllllllll) [inline zipline/pipeline/term.py]: we could use pandas strings like `M` for month end and `MS` for month start

Comment (twiecki) [inline zipline/pipeline/term.py]: +1 for using pandas' standard.

File: zipline/testing/fixtures.py [hunk 3]
```diff
@@ -527,7 +527,7 @@ def init_class_fixtures(cls):
         cls.sim_params = cls.make_simparams()
 
 
-class WithTradingSessions(WithTradingCalendars):
+class WithTradingSessions(WithTradingCalendars, WithDefaultDateBounds):
```
Comment (dmichalowicz) [inline zipline/testing/fixtures.py]: Looks like the order of these parent classes needs to be switched

File: zipline/pipeline/graph.py [hunk 4]
```diff
@@ -34,21 +33,23 @@ class TermGraph(DiGraph):
...
     ordered()
         Return a topologically-sorted iterator over the terms in self.
+
+    See Also
+    --------
+    ExecutionPlan
     """
     def __init__(self, terms):
-        super(TermGraph, self).__init__(self)
+        super(TermGraph, self).__init__()
```
Comment (dmichalowicz) [inline zipline/pipeline/graph.py]: Why this change?

File: zipline/pipeline/factors/factor.py [hunk 5]
```diff
@@ -1468,7 +1473,9 @@ def __iter__(self):
 
 
 class RecarrayField(SingleInputMixin, Factor):
-
+    """
+    A single field from a multi-output factor.
+    """
```
Comment (dmichalowicz) [inline zipline/pipeline/factors/factor.py]: 🎉

File: zipline/pipeline/graph.py [hunk 6]
```diff
@@ -58,6 +59,149 @@ def __init__(self, terms):
...
+    often needed when a term is an input to a rolling window computation.  For
+    example, if we compute a 30 day moving average of price from day X to day
+    Y, we need to load price data for the range from day (X - 29) to day Y.
+
+    Parameters
+    ----------
+    terms : dict
+        A dict mapping names to final output terms.
+    all_dates : pd.DatetimeIndex
+        The dates fo
```
Comment (dmichalowicz) [inline zipline/pipeline/graph.py]: Incomplete sentence ;)

File: zipline/testing/fixtures.py (hunk 3 above)
Comment (ssanderson) [inline zipline/testing/fixtures.py]: The order here shouldn't matter; both `WithTradingCalendars` and `WithDefaultDateBounds` are direct subclasses of `object`.  The tests are currently failing because another class is manually inheriting from both `WithTradingSessions` and `WithDefaultDateBounds`.  The right fix is to remove the duplication at that use-site.

File: zipline/pipeline/term.py (hunk 2 above)
Comment (ssanderson) [inline zipline/pipeline/term.py]: My only objection to this is that pandas inexplicably users `A` and `AS` for yearly instead of `Y` and `YS`, which I have never once successfully remembered.

File: zipline/pipeline/graph.py (hunk 4 above)
Comment (ssanderson) [inline zipline/pipeline/graph.py]: This was just a bug.  We were passing `self` twice to the parent constructor, which happened to work because the constructor of `DiGraph` takes another graph as its first argument, which is copied.  In this case we were forwarding an empty graph to the parent.

File: zipline/pipeline/term.py (hunk 2 above)
Comment (ssanderson) [inline zipline/pipeline/term.py]: @twiecki @llllllllll my other concern with using the pandas conventions is that I think calculating at the start of a period will be the more common desired outcome, but the more obvious period names of `'M'`, `'Y'` and `'Q'` mean all mean the end of the period.  Pandas also doesn't have a start/end distinction for weekly frequency.  Given that I think relatively few of our users are going to know the pandas symbols by heart, I don't think there's a ton to be gained by copying their conventions.  At the same time, there are probably enough users who are familiar with pandas that they'd be confused if we re-used `'M'` and `'Q'` but had them mean different things.

In light of all of the above, what do you think about just using slightly-more-verbose strings like `'month_start'`, `'year_start'`, `'quarter_end'`?

Comment (llllllllll) [inline zipline/pipeline/term.py]: I agree that we should not make it close but slightly different to pandas. If you really don't like the pandas version then I think the verbose names are fine. I think people will look this up when they need to use it anyways so following the pandas convention can help the few people that do know them.

File: zipline/pipeline/pipeline.py [hunk 7]
```diff
@@ -148,24 +148,57 @@ def set_screen(self, screen, overwrite=False):
...
+        Compile into a simple TermGraph with no extra row metadata.
+
+        Parameters
+        ----------
+        screen_name : str
+            Name to supply for self.screen.
+        default_screen : zipline.pipeline.term.Term
+            Term to use as a screen if self.screen is None.
+        """
+        return TermGraph(self._prepare_graph_terms())
```
Comment (dmichalowicz) [inline zipline/pipeline/pipeline.py]: I think this is missing its arguments

Comment (ssanderson) [inline zipline/pipeline/pipeline.py]: Good catch.  The codepath that calls this isn't tested because it relies on having the `dot` program installed.  I'll see if I can at least throw in a smoketest for this.

File: zipline/pipeline/engine.py [hunk 8]
```diff
@@ -170,7 +179,10 @@ def run_pipeline(self, pipeline, start_date, end_date):
             graph,
             dates,
             assets,
-            initial_workspace={self._root_mask_term: root_mask_values},
+            initial_workspace={
+                self._root_mask_term: root_mask_values,
+                self._root_mask_dates_term: dates.values[:, None],
```
Comment (dmichalowicz) [inline zipline/pipeline/engine.py]: Maybe use `as_column` here?

File: zipline/utils/numpy_utils.py [hunk 9]
```diff
@@ -364,3 +367,66 @@ def vectorized_is_element(array, choices):
...
+    """
+    if a.ndim != 1:
+        raise ValueError(
+            "as_column expected an 1-dimensional array, "
+            "but got an array of shape %s" % a.shape
+        )
+    return a[:, None]
+
+
+def changed_locations(a, include_first):
```
Comment (dmichalowicz) [inline zipline/utils/numpy_utils.py]: Do you think this is worth adding a short test for?

**Comment by dmichalowicz** [comment]
@ssanderson Done with a first review. I only skimmed through most of the tests though, so if there were certain parts of the tests you think I should look at more in depth look just let me know.

File: zipline/utils/numpy_utils.py (hunk 9 above)
Comment (ssanderson) [inline zipline/utils/numpy_utils.py]: The example here is run automatically by doctest.  Do you think it needs something beyond that?

File: zipline/pipeline/engine.py (hunk 8 above)
Comment (ssanderson) [inline zipline/pipeline/engine.py]: I think this is what I actually wrote `as_column` for and then forgot.  Will update.

File: zipline/utils/numpy_utils.py (hunk 9 above)
Comment (dmichalowicz) [inline zipline/utils/numpy_utils.py]: No I think that is sufficient (I did not know that)

File: zipline/pipeline/term.py (hunk 2 above)
Comment (twiecki) [inline zipline/pipeline/term.py]: I only knew the `BMonthStart` and `BMonthEnd` date offsets to begin with. The verbose names sound fine to me as well.

File: zipline/pipeline/mixins.py [hunk 10]
```diff
@@ -400,14 +400,35 @@ def _compute(self, windows, dates, assets, mask):
...
+                    next(w)
+        else:
+            # If we're not windowed, inputs are just ndarrays.  We need to
+            # slice off one row when forwarding to real_compute, but we don't
+            # need to do anything to skip an input.
+            def prepare_inputs():
+                # i is the loop iteration variable below.
+                return [a[[i]] for a in inputs]
+
+            def skip_this_input():
```
Comment (dmichalowicz) [inline zipline/pipeline/mixins.py]: Would this be more aptly named `skip_this_date`?

Comment (ssanderson) [inline zipline/pipeline/mixins.py]: I would think that that would encompass the whole action that happens in the loop iteration, whereas this just applies to the part that interacts with `inputs`.

File: zipline/pipeline/factors/factor.py [hunk 11]
```diff
@@ -1072,9 +1073,9 @@ def isfinite(self):
         """
         return (-inf < self) & (self < inf)
 
-    @property
+    @classlazyval
     def _downsampled_type(self):
-        return DownsampledFactor
+        return DownsampledMixin.make_downsampled_type(Factor)
```
Comment (dmichalowicz) [inline zipline/pipeline/factors/factor.py]: Can you remove `DownsampledFactor` now?

File: zipline/pipeline/mixins.py [hunk 12]
```diff
@@ -218,3 +238,224 @@ def _validate(self):
...
+        Factory for making Downsampled{Filter,Factor,Classifier}.
+        """
+        docstring = dedent(
+            """
+            A {t} that defers to another {t} at lower-than-daily frequency.
+
+            Parameters
+            ----------
+            term : {t}
+            {{frequency}}
```
Comment (dmichalowicz) [inline zipline/pipeline/mixins.py]: Does this require the `templated_docstring` decorator?

Comment (dmichalowicz) [inline zipline/pipeline/mixins.py]: Ah never mind, didn't see below

**Comment by dmichalowicz** [comment]
Other than the `to_simple_graph` fix, looks good to me
